[dev]: Refactor Checkout

sirugh opened the pull request:
## Description
Refactoring of checkout components, including updating unit tests. Notably: 

* Removed `SubmitButton` in favor of just using the `Button` component directly.
* Removed the unused `ResetButton`.
* Address Form Refactor
  * fixed bug in checkout reducer where street address was not saved
* BraintreeDropin Refactor
  * The conversion to the mount/unmount logic required some...interesting code. Check out the `useEffect` in the component and see how I had to use a local variable to ensure we teardown/error properly.
* Payment Form Refactor
  * identified a "bug" where `informed` updates state if `initialValues` are used causing multiple renders.
* Form Refactor
  * The parent form is now split into multiple function components. Should we move these into a `Form/` directory and split them out into individual files?


## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here by replacing ISSUE_NUMBER with your actual issue number. -->
<!--- Using the above wording causes Github to automatically close the issue on merge. -->
Closes #1294.

## Verification Steps
<!-- Please describe in detail how a reviewer can verify your changes. -->
1. Go to the checkout form and step through it. Everything should work including all validation, etc.
2. Try doing the checkout form out of order.
3. Try quickly opening and closing sub-forms such as the payments form.
4. Make sure form state is used appropriately. Does it remember address? Does it remember last choices such as "use same as shipping address", etc?

## How have YOU tested this?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Lots of manual testing through different scenarios. See above.

## Screenshots / Screen Captures (if appropriate):

## Proposed Labels for Change Type/Package
<!--- What type of change level would you suggest for this PR? -->
<!--- Major, Minor, or Patch? -->
<!--- See https://magento-research.github.io/pwa-studio/technologies/versioning/ for help -->
- [x] major (e.g x.0.0 - a breaking change)
- [ ] minor (e.g 0.x.0 - a backwards compatible addition)
- [ ] patch (e.g 0.0.x - a bug fix)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have updated the documentation accordingly, if necessary.
- [ ] I have added tests to cover my changes, if necessary.

sirugh commented on packages/venia-concept/src/components/Checkout/paymentsForm.js:
```diff
@@ -22,228 +21,232 @@ const DEFAULT_FORM_VALUES = {
+        submit,
+        submitting
+    } = props;
+    const formState = useFormState();
```
A wild hook appears!

jimbo replied:
Good comment. I'm not surprised, though; Informed implements change publishing manually, so they're not doing hooks + context in any sort of optimal way.

sirugh commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,354 @@
+    const handleSubmitAddressForm = useCallback(
+        formValues => {
+            submitShippingAddress({
+                type: 'shippingAddress',
```
I don't think we need to forward these `type` props since we have specific functions for submission of each type. Maybe a remnant from some earlier code? I see two ways to handle this, and I'd like opinions:

1) Remove `type`. Easiest "fix".
2) Refactor into a single `submitForm` action that accepts a `type` and handles logic internally instead of having multiple different `submit` functions. Benefit of only having to thread one function prop down instead of three.

sirugh replied:
I opted for option 1 right now since nothing is using `type`. And we can easily refactor the actions later.

sirugh commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,354 @@
-    }
+const isSubmitDisabled = (busy, valid) => busy || !valid;
+
+const EditableForm = props => {
```
The parent form is now split into multiple function components. Should we move these into a Form/ directory and split them out into individual files?

sirugh replied:
@jimbo says he prefers flat structure. I like nested, for things that are specific to a parent and won't be shared. 🤷‍♂

supernova-at replied:
> The parent form is now split into multiple function components. Should we move these into a Form/ directory and split them out into individual files?

We should definitely split them out into individual files, as per existing convention. I do like the idea of a `Form/` directory personally - especially since we have a `Receipt` directory within `Checkout` already.

supernova-at replied:
> @jimbo says he prefers flat structure. I like nested, for things that are specific to a parent and won't be shared. 🤷‍♂

![fight](https://user-images.githubusercontent.com/13182778/59468074-d108d900-8df6-11e9-8442-58ba0df40685.gif)

jimbo replied:
We should never have had a `Receipt` folder either—last year we had a few contributors who didn't like to follow conventions. 😉 

Honestly though, if you guys prefer to nest folders in places with a lot of files like this, I'm fine with it. Just don't go overboard.

sirugh commented on packages/venia-concept/src/components/Checkout/container.js:
```diff
@@ -0,0 +1,169 @@
+
+// TODO: This is exported for testing, but I think we should split this into a
+// separate file entirely and just do the prop/action shaping here in this file.
+export const CheckoutContainer = props => {
```
Yea, I'm not even sure we need this component. We do the prop shaping, split it into multiple groups, and then [pull them apart](https://github.com/magento-research/pwa-studio/pull/1309/files#diff-c54588d5f498544c2028e38a1865dc0aR48) inside flow anyways.

sirugh commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,366 @@
+            await submitShippingAddress({
+                formValues
+            });
+            setEditing(null);
```
If we move state out of redux we must at some point `await` some async action completion where the logic would have [otherwise been handled internally in the reducer](https://github.com/magento-research/pwa-studio/blob/develop/packages/venia-concept/src/reducers/checkout.js#L116). For example, we would have `setEditing(null)` in the `paymentMethod.accept` function instead. We will see the same sort of logic when/if we convert the checkout `step` to a local state.

Is there some best practice when it comes to doing this sort of thing? Personally I think this reads better but it has a consequence of "hiding" these state changes within the components.

sirugh replied:
After discussion with @jimbo we decided that while this may be the right direction to take when it comes to refactoring out redux, we don't want to do it _here_. So I will be reverting this commit into a separate branch/PR so we have it around (since I spent time on updating tests). And we can deal with it later.

sirugh replied:
Moved to #1338

supernova-at commented on packages/venia-concept/src/components/Checkout/cart.js:
```diff
@@ -1,35 +1,34 @@
-                        submit={beginCheckout}
-                    />
-                </div>
+    const disabled = isDisabled(submitting, ready);
```
Do we use `isDisabled` anywhere else? Want to just inline its logic here?

sirugh replied:
Sure. This used to live in `CheckoutButton` but moving it here allowed me to simplify the button API, which IMO isn't super necessary anyways. I may end up replacing it entirely with one of the shared button/trigger components.

supernova-at commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,364 @@
+            return null;
+        }
+    }
+};
```
Would be nice to have `propTypes` defined for each of these subcomponents.

sirugh replied:
WOULDN'T IT!?



Fine :D

jimbo replied:
And put them in new files! 💯

sirugh replied:
DONE. 💀

supernova-at commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,364 @@
+                    submit={handleSubmitAddressForm}
+                    isAddressIncorrect={isAddressIncorrect}
+                    incorrectAddressMessage={incorrectAddressMessage}
+                />
```
This wouldn't block a 👍 but just pointing out that alphabetizing really helps readability in cases where there are a lot of props like this.

sirugh replied:
Definitely. I did it in some places but not others. I shall look for and alphabetize the props on component implementations :)

sirugh replied:
I think I'm tired of staring at this stuff. I did it where I could. If we want to enforce this we should add some linter/plugin that alphabetizes this for us automatically. Ugh. lol

supernova-at commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,364 @@
+        default: {
+            return null;
+        }
+    }
```
This looks great!

jimbo replied:
Agreed. Much cleaner. 👍

supernova-at commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,364 @@
+    submitting: bool.isRequired
+};
+
+export default Form;
```
Man, all this is so much easier to reason about now. Great work!

sirugh commented on packages/venia-concept/src/components/Checkout/flow.js:
```diff
@@ -1,192 +1,157 @@
+
+    return [billingAddress, paymentData, shippingAddress, shippingMethod].every(
+        hasData
+    );
```
Good point. I didn't update these util functions, just move them, but I will update the check.

sirugh replied:
So all of these props start as null. As they are filled they are added to the store. All are added as objects except for `shippingMethod` which is a string. I can update this check to be more explicit.

sirugh commented on packages/venia-concept/src/components/Checkout/flow.js:
```diff
@@ -1,192 +1,157 @@
+
+    // ensure state slices are present
+    if (!(cart && checkout)) {
+        return null;
```
TBH I don't know why this is necessary. We shouldn't ever be setting these to null as they are state objects that come from redux.

jimbo replied:
This is legacy code from when these slices were dynamically imported and added to the root reducer. We don't do that anymore, so the safeguard should no longer be necessary

sirugh commented on packages/venia-concept/src/components/Checkout/flow.js:
```diff
@@ -1,192 +1,157 @@
+    cancelCheckout: func,
+    cart: shape({
+        details: object.isRequired
+    }),
```
Yea, I was moving props around when I removed the container. `Flow` didn't previously have the helper function as it was in `wrapper`, and that had proptypes which required `cart` and `details`. I think I ran into a problem where `cart` was an empty object and was failing the prop checking. I'll investigate.

jimbo replied:
@supernova-at Good analysis.

jimbo commented on packages/venia-concept/src/components/Checkout/paymentsForm.js:
```diff
@@ -22,228 +21,229 @@ const DEFAULT_FORM_VALUES = {
+            <div className={classes.street0}>
+                <Field label="Street">
+                    <TextInput
+                        id={classes.street0}
```
Not that this is your problem (in fact, this might be my code), but the `id` attribute may not contain spaces. The classnames output by CSS Modules may contain spaces, though, so I don't think we can use them like this.

sirugh replied:
I don't think I follow. Can you clarify @jimbo?

jimbo commented on packages/venia-concept/src/components/Checkout/paymentsForm.js:
```diff
@@ -22,228 +21,229 @@ const DEFAULT_FORM_VALUES = {
-    render() {
-        const { classes, initialValues } = this.props;
-        const { formChildren } = this;
+const BillingAddressFields = ({ classes, countries }) => {
```
We definitely want to stick to one component definition per file. There's nothing wrong with having more files.

jimbo commented on packages/venia-concept/src/components/Checkout/paymentsForm.js:
```diff
@@ -22,228 +21,229 @@ const DEFAULT_FORM_VALUES = {
+    // The success callback. Unfortunately since form state is created first and
+    // then modified when using initialValues any component who uses this
+    // callback will be rendered multiple times on first render. See above
+    // comments for more info.
```
Unfortunately, that's the reality of almost any convenient API where we set initial values, validators, etc., on individual components. Those things have to be sent up to the form's state object for actual initialization, meaning it takes a re-render before everything is fully set up.

I think that re-render is acceptable, and it'll be even better if we ever move away from Informed.

jimbo commented on packages/venia-concept/src/components/Checkout/paymentsForm.js:
```diff
@@ -22,228 +21,229 @@ const DEFAULT_FORM_VALUES = {
+            });
+        },
+        [formState.values]
+    );
```
Good work on these callbacks. 👍

jimbo commented on packages/venia-concept/src/components/Checkout/paymentsForm.js:
```diff
@@ -22,228 +21,229 @@ const DEFAULT_FORM_VALUES = {
+                        shouldRequestPaymentNonce={isSubmitting}
+                        onError={handleError}
+                        onSuccess={handleSuccess}
+                    />
```
Is there any way for `BraintreeDropin` to be loaded with `React.lazy()` and `Suspense`?

sirugh replied:
I can investigate. TBH I haven't used either yet, so I'm eager to try.

sirugh replied:
Do you still want me to look into this? Not sure how much work it'll be do we want to add to this PR?

jimbo commented on packages/venia-concept/src/components/Checkout/label.js:
```diff
@@ -1,25 +1,27 @@
+const Label = props => {
+    const { children, plain, ...restProps } = props;
+    const classes = mergeClasses(defaultClasses, props.classes);
+    const elementType = plain ? 'span' : 'label';
```
Thanks for keeping this. There are no comments in the file, but it's important not to have nested `<label>` elements for accessibility.

jimbo commented on packages/venia-concept/src/components/Checkout/form.js:
```diff
@@ -1,329 +1,364 @@
+ */
+const Form = props => {
+    const classes = mergeClasses(defaultClasses, props.classes);
+    let child;
```
Not that it's *super* important, but you should be able to use `const` even in places like this. We could tighten up the `prefer-const` rule if warnings would help.

```es6
const child = editing ? (
  <Foo />
) : (
  <Bar />
)

```

jimbo commented on packages/venia-concept/src/components/Checkout/braintreeDropin.js:
```diff
@@ -1,139 +1,147 @@
+
+        // If we close the form quickly after opening we may end up in a
+        // semi-mounted state so we need a local variable to make sure we
+        // behave correctly.
```
Great work. This is obviously important to handle but easy to overlook.

jimbo commented on packages/venia-concept/src/components/Checkout/braintreeDropin.js:
```diff
@@ -1,139 +1,147 @@
+ *
+ * 1) Mount and asynchronously create the dropin via the braintree api.
+ * 2) On submission (triggered by a parent), request the payment nonce.
+ */
```
This component is a great example of what we're aiming for. 👍

jimbo commented on packages/venia-concept/src/components/Checkout/addressForm.js:
```diff
@@ -27,180 +26,175 @@ const fields = [
-    static defaultProps = {
-        initialValues: {}
-    };
+    const values = useMemo(
```
Good use of `useMemo()` here.

sirugh commented on packages/venia-concept/src/components/Checkout/section.js:
```diff
@@ -1,54 +1,48 @@
+        summary: string
+    }),
+    label: node,
+    onClick: func,
```
I don't remember 😢

sirugh commented on packages/venia-concept/src/middleware/log.js:
```diff
@@ -10,13 +10,17 @@ const log = store => next => action => {
+    // logged as empty even though they aren't and the state is updated with
+    // incoming values (although that is dependent on the action). Probably
+    // related to Chrome logging a reference to the array and some code clearing
+    // the array at some point.
```
Ok.

sirugh commented:
I addressed most of the feedback. There's a few things I need to look into such as lazy loading braintree and some leftover TODOs. I also split out `Form` into multiple component files. I chose not to create a `Form/` folder right now because I didn't want to lose some of the requested changes. I can do that later if we care, otherwise I'm going to leave it as is.

supernova-at commented on packages/venia-concept/src/components/Checkout/braintreeDropin.js:
```diff
@@ -1,139 +1,147 @@
+    shouldRequestPaymentNonce: bool
+};
+
+export default BraintreeDropin;
```
Looks great!

sirugh commented on packages/venia-concept/src/components/Checkout/addressForm.js:
```diff
@@ -27,180 +26,176 @@ const fields = [
-    };
+    const handleCancel = useCallback(() => {
+        cancel();
+    }, [cancel]);
```
Yea I had this same feedback given on another PR. The only reason for renaming the callback would be so that it falls into some paradigm we use throughout the application. Of course, it's technically overkill and unnecessary. I am totally fine removing this sort of use of `useCallback` with a direct passing of the prop.

jimbo commented on packages/venia-concept/src/components/Checkout/paymentsForm.js:
```diff
@@ -21,8 +21,26 @@ const DEFAULT_FORM_VALUES = {
 
-const BillingAddressFields = ({ classes, countries }) => {
-    return (
+const PaymentsFormChildren = props => {
```
`PaymentsFormChildren` should also be extracted into its own file, no?

Also, consider `PaymentsFormItems`, if you'd like to follow the pattern we have in a couple other places.

sirugh replied:
I wasn't sure about this -- the component wrapping it is literally just one layer:

```
return (
  <div>
    <PaymentsFormChildren {...props} />
   </div>
)
```
We can probably just get rid of this separated component.

sirugh replied:
Oh I see -- it's pretty involved. Maybe I need to break out the `BillingAddressFields` too...

sirugh commented:
@jimbo feedback addressed! If this is good to go please put it in Ready for QA :)
